### ISSUE TITLE
Indirect package manager launch through Command

### ISSUE THREAD

**Issue by shidel** [shidel/FDIMPLES#5]
Hi @ecm-pushbx, 

FDIMPLES has a compiler directive _USEBAT_ to use COMMAND.COM as a stop-gap solution for the issue that happens when trying to execute FDINST.COM directly.  So, I've just compiled the most recent version with that disabled. 

It is very easy to replicate, simply run FDIMPLES, pick any package and change its state.  Either To install or To Remove. It does not matter which way. Then, tab down to OK and hit enter. 

![Screen Shot 2024-05-05 at 4 55 42 PM](https://github.com/shidel/FDIMPLES/assets/10198034/bb58b7ab-f2f7-406d-adcb-6c3e9247d4ce)

FDINST replies that "(blank) is not installed"  and exits with an ERRORLEVEL. 

When _USEBAT_ is enabled, it creates a small batch with the same commands it would pass directly. then executes COMMAN.COM passing the batch file as a parameter to execute. It's a very ugly solution. But, it worked to get past the issue in launching FDINST directly. Other programs seem to have no problem being directly launched. The PGME (menu based program launcher)  uses  my same code to execute programs and has the same problem with FDINST. 

Attached is the PASCAL code from that contains the EXEC routines (around line 980) and a version of FDIMPLES compiled without _USEBAT_ enabled. 

[FDIMPLES.zip](https://github.com/shidel/FDIMPLES/files/15214589/FDIMPLES.zip)
[QDOS.zip](https://github.com/shidel/FDIMPLES/files/15214592/QDOS.zip)

**Comment by boeckmann** [comment]
FDINST.COM source is at https://gitlab.com/FreeDOS/util/fdnpkg/-/tree/master/SOURCE/FDNPKG

**Comment by boeckmann** [comment]
[Source ](https://gitlab.com/FreeDOS/util/fdnpkg/-/blob/master/SOURCE/FDNPKG/PKGREM.C#L88 )indicates that the complete error message should be:

    kitten_printf(4, 0, "Package %s is not installed, so not removed.", pkgname);

`Package ` is missing in the screenshot! Perhaps a terminating carriage return of a malformed command line / environment is causing trouble? I think we should substitute FDINST by a stub that prints the command line and environment variables to see if that is ok.

**Comment by shidel** [comment]
I just made a small tweak to the Exec routine. It outputs the strings passed to the DOS Function call. As **(INT VALUE)ASCII**. the ASCII for CR & LF are not displayed. 

![Screen Shot 2024-05-05 at 5 39 06 PM](https://github.com/shidel/FDIMPLES/assets/10198034/826ea84a-49e9-4182-97ac-56073306b2b7)

**Comment by boeckmann** [comment]
Isn't the last (13) a carriage return?

**Comment by shidel** [comment]
> Isn't the last (13) a carriage return?

Yes, intentionally. 
Although I do not recall which at the moment, many programs do like it if that is not at the end of the line.

**Comment by boeckmann** [comment]
Perhaps FDINST considers the CR as part of its second argument? That would explain why a) it does not find the file and b) it swallows the first part of the error message.

**Comment by shidel** [comment]
> Perhaps FDINST considers the CR as part of its second argument? That would explain why a) it does not find the file and b) it swallows the first part of the error message.

Nope. 

Since FDINST is the only program FDIMPLES executes. I've tried removing it and just NULL terminating the string. Same result. I've also added a 0 after the CR. No change.

I've even tried putting a extra space in front of things. Nope. 

It's weird. But, as I sit here now. I think I may have an idea...

It may be that FDINST assumes the parameter string immediately follows the command string in memory. Which is not the case and is not required by the spec (as far as I know). You provide two pointers to DOS to perform the Exec. One as the PATH\COMMAND and the other to PARAMETERS. But DOS only populates the PSP with the parameter string. So, ?

Actually, it wasn't until recently that I figured out how to get the PATH\COMMAND that was used to launch a program from inside a COM binary. Doing so from an EXE is easy. But, doing it from a COM is weird and not well known at all. It's something I figured out while working on V8TURBO.

**Comment by shidel** [comment]
It's been so long.  I forgot. 

But, looking at the ASCII used to launch FDINST...

FDINST is an EXE not a COM. 

So, I wonder if that is what is happening.

**Comment by boeckmann** [comment]
At least FDINST seems to accept its first argument "remove". Otherwise it would not try to remove :) It's after midnight in Aachen. Tomorrow I will substitute FDINST by a stub and have a look at the program arguments from a client side.

**Comment by shidel** [comment]
Sounds good. 
Just remember to use the version in this thread. Or compile it with the _USEBAT_ directive disabled. Otherwise, it will pass it through command.com.

**Comment by boeckmann** [comment]
Still problems? Under T2405 running in 86box its running fine for me.

**Comment by shidel** [comment]
Hi @boeckmann, 

I reopened this because I was doing some testing and noticed something odd. Well, actually a couple of odd things.

First, I slapped together a quick program to display the command line _(in color using the new library stuff)_.

Then I was testing really long command lines sent to the program. Here are the results...

With the parameters at 125 _(0x7d)_ characters, _FreeDOS_ appends a **CR**+**NULL**. Great.

![Screenshot_20240508_063605](https://github.com/shidel/FDIMPLES/assets/10198034/fb0c2227-31aa-4be3-8bdd-6a388efdb58c)

At 126 _(0x7e)_ characters, _FreeDOS_ drops the **NULL** and just appends the **CR**. Great.

![Screenshot_20240508_063815](https://github.com/shidel/FDIMPLES/assets/10198034/20eca2ee-d840-405f-8216-ca791cf67482)

At 127 _(0x7f)_ characters, _FreeDOS_ truncates the string to 126 characters and places a **CR** at the end. Great. But at 127 characters and beyond the length is left at **0x7f**. 

![Screenshot_20240508_063839](https://github.com/shidel/FDIMPLES/assets/10198034/8ab713d8-e841-4fa6-b116-ad22332ffe3a)

Maybe it is fine. But, this does not seem correct to me.  

If a program _(as discussed early)_ uses the length byte to calculate the end of the string and then overwrite the **CR** with a **NULL**, then the **NULL** would be written to **0x0100**. Since, the instruction at that memory location would have already been executed, that is probably ok. Unless that program loops or jumps back to **0x0100** for some reason. The problem is that the **CR** would still be left in the string. 

So, I decided to see what happens under _MS-DOS_, _PC-DOS_ and _DR-DOS_.

I copied the test program onto an SD-CARD and booted it in the Pentium Pro under _MS-DOS 6.22_. Ran the program, got an `insufficient memory` error _(more on this bellow)_. So, I made a quick change and repeated the process. 

Under _MS-DOS 6.22_ and _PC-DOS 7.10_, you can not enter more than 127 characters on the command line. This includes whatever program you are executing. It has been so long, I forgot about that. So, renaming the program to just M.COM allowed the entire use of the parameter string. Both only append a **CR** and have a maximum length of parameter string to 126 _(0x7e)_ characters. 

Under _DR-DOS 7.03_, you can only enter 126 characters on the command line. It always appends **CR**+**NULL**. Limiting the maximum length to 125 _(0x7d)_ characters for parameters.

So none of the others exceed 126 _(0x7e)_ for the maximum length from the command line. However, I did not try from a BATCH file. Perhaps that may exhibit different results. 

Now on the the `insufficient memory` error. 

At startup the library always verifies the user is running at least DOS 5.0. It also verifies that the entire segment for the program is available for use.  Written under _DOSBox_+_FreeCOM_, the program simply tests the **CS:0006** _(last byte in first segment of program)_ is **0xffff**. But, apparently this is a _DOSBox_ bug. Because all other DOS kernels tested, _FreeDOS_, _DR-DOS_, _MS-DOS_ and _PC-DOS_ all set this value to **0xfef0**. The logic behind this is obvious. The program stack is up there initially and they are "reserving" some space to help protect it. So, the memory test will possibly switch to requiring **0xfef0** or higher. Or, more likely just check that **SP** is **0xfffe**. Because, all DOS kernels push a **0x0000** _word_ onto the stack for the program to be able to just **RET** to exit. Not generally advisable to actually terminate a program that way. But, it is there.

[ME.ZIP](https://github.com/shidel/FDIMPLES/files/15248894/ME.ZIP)

**Comment by shidel** [comment]
> Still problems? Under T2405 running in 86box its running fine for me.

FDIMPLES seems good now. I just reopened it because of the weirdness I noticed during some testing of the new library stuff.

**Comment by boeckmann** [comment]
I tested it with FreeCOM. It accepts a (if I counted correctly) 254 character input at the command prompt. That vastly exceeds the 127 bytes available for the program arguments in the PSP. Would be interesting to know, why the FreeCOM authors allowed this. It may be to able entering a long executable path in addition to the maximum 127 bytes of the program arguments.

I consider it to be dangerous, especially if the command line gets shortened automatically. Consider you are trying to delete a file, and the name changes because of the truncation. Chances are low but there may also be a file under the shortened file name. I'd have to dig into FreeCOM to gather further insight. But that would probably better placed as an issue in the FreeCOM repo.

**Comment by boeckmann** [comment]
Additional detail: FreeCOM internal commands seem to be able to process the full 254 bytes of the command line. Running an externel program, the arguments get shortened to 127 bytes. I think it would be appropriate to raise an error in this case instead of executing the command.

**Comment by shidel** [comment]
> I tested it with FreeCOM. It accepts a (if I counted correctly) 254 character input at the command prompt. That vastly exceeds the 127 bytes available for the program arguments in the PSP. Would be interesting to know, why the FreeCOM authors allowed this. It may be to able entering a long executable path in addition to the maximum 127 bytes of the program arguments.

Well, there are a couple reasons it accepts more than 127 bytes.

First I assume, because it could have a long path to the executable plus a full set parameters.

It should probably be a lot larger when LFN support is available. 

Unlike the other DOS variants, when the parameters exceed the PSP buffer, it sets an Environment variable called CMDLINE. I’ve not used that feature before. So, I don’t recall if it is the full string or just the overflow. Most likely, the full string.

I have to finish a couple simple macros for the library before I look at that. Just very busy ATM.

**Comment by shidel** [comment]
Ok... After some head banging, the simple macros I needed were written and I added them to ME.COM to test some stuff regarding the command line parameters passed to the program. 

Now, when there are 0x7f characters or more, FreeCOM passes the entire command line as an Environment Variable called "CMDLINE".  

While I was change the length up and down of the parameters, I though I saw a truncation bug in the string passed through CMDLINE. However, I was unable to duplicate it. It is possible it was a glitch in the FreeCOM command line editor and not the passing of the parameter. 

![Screenshot_20240510_011253](https://github.com/shidel/FDIMPLES/assets/10198034/92d3cb1e-547a-4b7e-957a-d45a6146a432)

The issues with command line editor in FreeCOM are fairly easy to reproduce. Just have a multiple lines, move towards the beginning of the text, use the up/down history keys. Or, at the beginning of the text, delete a few characters and start typing new ones. There are several quirks in the command line editor. 

[ME.COM.zip](https://github.com/shidel/FDIMPLES/files/15270523/ME.COM.zip)
